### ISSUE TITLE
External JSTransformer Integration

### ISSUE THREAD

**Issue by RobLoach** [jstransformers/jstransformer#17]
I put together a [Metalsmith plugin to use JSTransformer](http://github.com/robloach/metalsmith-jstransformer), and thought it would be a good idea to have a list of potential ideas for other integration points with external systems. Feel free to edit this list and add any other ideas.

- [ ] [Jade](http://jade-lang.com) - Put together a [Jade filter](http://jade-lang.com/reference/filters/)
- [x] [Metalsmith](http://metalsmith.io) - [`metalsmith-jstransformer`](https://github.com/robloach/metalsmith-jstransformer)
- [ ] [Grunt](http://gruntjs.com/) - Plugin to process files through JSTransformers
- [ ] [Gulp](http://gulpjs.com) - Plugin to process streams through JSTransformers

Should some of these live in the `jstransformers` org?

**Comment by tunnckoCore** [comment]
> Should some of these live in the jstransformers org?

yea, I think it should.

**Comment by tunnckoCore** [comment]
Yesterday I thought for something like `jstransformer-duo` jstransformer-gulp`, but Im not sure what is my exact idea, lol.

But your ideas SG2M and easy to be done.

**Comment by tunnckoCore** [comment]
Also maybe as `koa` plugin? hmm. I'll think of it when free my mind.

**Comment by tunnckoCore** [comment]
btw, also just popped up in my mind, something called "jstemplates" or "jsengines" or somthing else (names welcome) which will be on top of jstransformer to normalize API for benchmarks.

**Comment by TimothyGu** [comment]
@RobLoach said:

> - [ ] [Jade](http://jade-lang.com) - Put together a [Jade filter](http://jade-lang.com/reference/filters/)

We are planning on completely switching to jstransformers in Jade, so:

```jade
:supermarked
  ## Heading 2
```

will be mapped to

```js
var name = 'supermarked'
require('jstransformers')(require(name)).render(text).body
```

or something like that.

@ForbesLindesay should be able to tell you more about Jade integration.

**Comment by ForbesLindesay** [comment]
Yes, jade will have full built in support for jstransformers.  See https://github.com/jadejs/jade/pull/1928 for approximately how this will be implemented.

**Comment by ForbesLindesay** [comment]
I've added browserify to the list.

**Comment by tunnckoCore** [comment]
> I've added browserify to the list. - @ForbesLindesay 

`jstransformify` is okey?

**Comment by calebeby** [comment]
@tunnckoCore https://github.com/andreypopp/jstransformify

**Comment by calebeby** [comment]
`jstransformerify`?

**Comment by tunnckoCore** [comment]
haha... probably :D

**Comment by tunnckoCore** [comment]
I just added `rollup` and `start` bullets.

**Comment by tunnckoCore** [comment]
https://github.com/tunnckoCore/start-jstransformer is started

**Comment by tunnckoCore** [comment]
Btw, reminder (at least for me). Rollup plugin would be awesome. Don't know if you thought for it, but we can use `inputformat-to-jstransformer` to load different kind of sources - won't be needed "loaders" and etc.. we already have huge database of "loaders" :)) So it would look like more awesome Webpack haha.

**Comment by calebeby** [comment]
:+1:

**Comment by tunnckoCore** [comment]
`rollup-plugin-jstransformer`:

```js
function jstransformer (opts) {
  const filter = createFilter(opts.include, opts.exclude, opts)

  return {
    name: 'jstransformer',
    transform (code, id) {
      if (!filter(id)) {
        return null
      }

      return new Promise((resolve) => {
        JSTransformer.renderAsync(code, {
          engine: opts.engine
        }, opts.locals).then((result) => {
          resolve({
            code: `export default ${JSON.stringify(result.body)};`,
            map: result.map || { mappings: '' }
          })
        })
      })
    }
  }
}
```

where `createFilter` can be from `rollup-pluginutils` or

```js
const path = require('path')
const micromatch = require('micromatch')

const arrayify = (val) => {
  if (!val) return []
  if (Array.isArray(val)) return val
  return [val]
}

const createFilter = (include, exclude, options) => (id) => {
  if (typeof id !== 'string') return false
  if (/\0/.test(id)) return false

  const fn = (pattern) => path.resolve(pattern)
  include = arrayify(include).map(fn)
  exclude = arrayify(exclude).map(fn)

  options = Object.assign({ ignore: exclude }, options)
  return micromatch(id, include, options).length > 0
}
```

it is working and uses micromatch better as v2 of pluginutils (suggested by me)

**usage**

```js
const jstransformer = require('rollup-plugin-jstransformer')
module.exports = {
  plugins: [
    jstransformer({
      engine: 'handlebars',
      locals: { adjective: 'cool' },
      include: './src/components/Test2.html'
    })
  ]
}
```

**Comment by calebeby** [comment]
:+1:
